cmd/broker/ingress/metrics.go

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+const (
+	metricsNamespace = "broker_ingress"
+)
+
+var (
+	// MeasureMessagesTotal is a counter which records the number of messages received
+	// by the ingress. The value of the Result tag indicates whether the message
+	// was filtered or dispatched.
+	MeasureMessagesTotal = stats.Int64(
```

Comment:
Can this metric have the broker namespace and name as a parameter? So that we can have the ability to distinguish the number of events that have gone through broker-a and broker-b. 

For ingress in particular, it is tied to a single broker, so we might be able to use some knowledge based on the Pod, but for triggers, we won't have that ability. The code will know which trigger a metric should be recorded against, but not the Pod.

I don't know if it is possible in OpenCensus, tags kind of look like what I am thinking about.

Reply:
Based on my reading of the Prometheus docs, namespace is normally added as a tag by its Kubernetes service discovery implementation: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#pod

Serving additionally specifies in its scrape config that the `__meta_kubernetes_namespace` label should be renamed to `namespace`:
https://github.com/knative/serving/blob/d454ec94aa1ec39d9f28539a6a0d6753c67f09ef/config/monitoring/metrics/prometheus/100-prometheus-scrape-config.yaml

Based on this prior art I think it's not necessary to specify the namespace tag, because the Pod namespace will always match. **However** if the broker ingress ever changes to be cluster-wide or work across namespaces, then metrics will need to be manually tagged with the Broker's namespace. In that case the implementation is likely to have access to the Broker object for generating tags.

I'll set the Broker name in an optional environment variable set by the controller.